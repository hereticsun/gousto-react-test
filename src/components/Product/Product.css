.product__title {
  background: none;
  border: none;
  font-size: 1rem;
  font-weight: normal;
  cursor: pointer;
}

.product__description {
  overflow: hidden;
  height: 0;
}

.product__description--expanded {
  height: auto;
}
