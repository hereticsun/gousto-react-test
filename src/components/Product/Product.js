import React, { Component } from 'react';
import PropTypes from 'prop-types';

import './Product.css';

export class Product extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isExpanded: false,
    };
  }

  toggleDescription(){
    this.setState({
      isExpanded: !this.state.isExpanded
    });
  };

  render() {
    const {product} = this.props;
    const isExpanded = this.state.isExpanded ? 'product__description--expanded' : '';
    return (
      <li
        key={product.id}
        className="product"
      >
        <button
          className="product__title" 
          style={{fontWeight: this.state.isExpanded ? 'bold' : null}}
          onClick={this.toggleDescription.bind(this)}>
          {product.title}
        </button>
        <div className={`product__description ${isExpanded}`}>
          <p>{product.description}</p>
        </div>
      </li>
    );
  }
}

Product.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.string,
    sku: PropTypes.string,
    title: PropTypes.string,
    description: PropTypes.string,
    list_price: PropTypes.string,
    is_vatable: PropTypes.bool,
    is_for_sale: PropTypes.bool,
    age_restricted: PropTypes.bool,
    box_limit: PropTypes.number,
    always_on_menu: PropTypes.bool,
    volume: PropTypes.number,
    zone: PropTypes.string,
    created_at: PropTypes.string,
    categories: PropTypes.arrayOf(PropTypes.object),
    tags: PropTypes.array,
    images: PropTypes.shape({
      365: PropTypes.shape({
        src: PropTypes.string,
        url: PropTypes.string,
        width: PropTypes.number,
      }),
    }),
  })
};

export default Product;
