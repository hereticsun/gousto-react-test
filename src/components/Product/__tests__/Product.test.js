import React from 'react';
import { shallow } from 'enzyme';

import {Product} from '../Product';

describe(`<Product />`, () => {
  describe('Given that I am a user', () => {
    const props = {
      product: {
        id: '1',
        title: 'Borsao Macabeo',
        description: 'A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.',
      }
    }

  describe('when I land on the main page', () => {
      const element = shallow(<Product {...props} />);

      it('displays the product', () => {
        expect(element).toMatchSnapshot();
        expect(element.find('.product__title')).toBeDefined();
        expect(element.find('.product__title').props().style).not.toEqual({fontWeight: 'bold'});
        expect(element.find('.product__description').hasClass('product__description--expanded')).toBeFalsy();
      });
    });

    describe('when I click on the product title', () => {
      const element = shallow(<Product {...props} />);
      it('displays the expanded product', () => {
        element.find('.product__title').simulate('click');
        expect(element).toMatchSnapshot();
        expect(element.find('.product__title').props().style).toEqual({fontWeight: 'bold'});
        expect(element.find('.product__description').hasClass('product__description--expanded')).toBeTruthy();
      });
    });

    describe('when I click on the product title a second time', () => {
      const element = shallow(<Product {...props} />);
      it('displays the collapsed product', () => {
        element.find('.product__title').simulate('click');
        element.find('.product__title').simulate('click');
        expect(element).toMatchSnapshot();
        expect(element.find('.product__title').props().style).not.toEqual({fontWeight: 'bold'});
        expect(element.find('.product__description').hasClass('product__description--expanded')).toBeFalsy();
      });
    });
  });
});
