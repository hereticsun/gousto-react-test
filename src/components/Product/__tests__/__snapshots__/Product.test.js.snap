// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> Given that I am a user when I click on the product title a second time displays the collapsed product 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Product
    product={
      Object {
        "description": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
        "id": "1",
        "title": "Borsao Macabeo",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "1",
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="product__title"
          onClick={[Function]}
          style={
            Object {
              "fontWeight": null,
            }
          }
        >
          Borsao Macabeo
        </button>,
        <div
          className="product__description "
        >
          <p>
            A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.
          </p>
        </div>,
      ],
      "className": "product",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Borsao Macabeo",
          "className": "product__title",
          "onClick": [Function],
          "style": Object {
            "fontWeight": null,
          },
        },
        "ref": null,
        "rendered": "Borsao Macabeo",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <p>
            A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.
          </p>,
          "className": "product__description ",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
          },
          "ref": null,
          "rendered": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
          "type": "p",
        },
        "type": "div",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "1",
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="product__title"
            onClick={[Function]}
            style={
              Object {
                "fontWeight": null,
              }
            }
          >
            Borsao Macabeo
          </button>,
          <div
            className="product__description "
          >
            <p>
              A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.
            </p>
          </div>,
        ],
        "className": "product",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Borsao Macabeo",
            "className": "product__title",
            "onClick": [Function],
            "style": Object {
              "fontWeight": null,
            },
          },
          "ref": null,
          "rendered": "Borsao Macabeo",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <p>
              A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.
            </p>,
            "className": "product__description ",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
            },
            "ref": null,
            "rendered": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
            "type": "p",
          },
          "type": "div",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`<Product /> Given that I am a user when I click on the product title displays the expanded product 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Product
    product={
      Object {
        "description": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
        "id": "1",
        "title": "Borsao Macabeo",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "1",
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="product__title"
          onClick={[Function]}
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          Borsao Macabeo
        </button>,
        <div
          className="product__description product__description--expanded"
        >
          <p>
            A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.
          </p>
        </div>,
      ],
      "className": "product",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Borsao Macabeo",
          "className": "product__title",
          "onClick": [Function],
          "style": Object {
            "fontWeight": "bold",
          },
        },
        "ref": null,
        "rendered": "Borsao Macabeo",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <p>
            A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.
          </p>,
          "className": "product__description product__description--expanded",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
          },
          "ref": null,
          "rendered": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
          "type": "p",
        },
        "type": "div",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "1",
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="product__title"
            onClick={[Function]}
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Borsao Macabeo
          </button>,
          <div
            className="product__description product__description--expanded"
          >
            <p>
              A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.
            </p>
          </div>,
        ],
        "className": "product",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Borsao Macabeo",
            "className": "product__title",
            "onClick": [Function],
            "style": Object {
              "fontWeight": "bold",
            },
          },
          "ref": null,
          "rendered": "Borsao Macabeo",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <p>
              A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.
            </p>,
            "className": "product__description product__description--expanded",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
            },
            "ref": null,
            "rendered": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
            "type": "p",
          },
          "type": "div",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`<Product /> Given that I am a user when I land on the main page displays the product 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Product
    product={
      Object {
        "description": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
        "id": "1",
        "title": "Borsao Macabeo",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "1",
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="product__title"
          onClick={[Function]}
          style={
            Object {
              "fontWeight": null,
            }
          }
        >
          Borsao Macabeo
        </button>,
        <div
          className="product__description "
        >
          <p>
            A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.
          </p>
        </div>,
      ],
      "className": "product",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Borsao Macabeo",
          "className": "product__title",
          "onClick": [Function],
          "style": Object {
            "fontWeight": null,
          },
        },
        "ref": null,
        "rendered": "Borsao Macabeo",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <p>
            A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.
          </p>,
          "className": "product__description ",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
          },
          "ref": null,
          "rendered": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
          "type": "p",
        },
        "type": "div",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "1",
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="product__title"
            onClick={[Function]}
            style={
              Object {
                "fontWeight": null,
              }
            }
          >
            Borsao Macabeo
          </button>,
          <div
            className="product__description "
          >
            <p>
              A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.
            </p>
          </div>,
        ],
        "className": "product",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Borsao Macabeo",
            "className": "product__title",
            "onClick": [Function],
            "style": Object {
              "fontWeight": null,
            },
          },
          "ref": null,
          "rendered": "Borsao Macabeo",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <p>
              A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.
            </p>,
            "className": "product__description ",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
            },
            "ref": null,
            "rendered": "A flavoursome Summer wine made from the indigenous Macabeo grape in northern Spain. A balanced, modern white with flavours of ripe peach, zesty lemon and nutty undertones, it leaves the palate with a clean and fruity finish.",
            "type": "p",
          },
          "type": "div",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
